File: Algorithm/python_source/charpter2_reverselist/level2/topic2_2两两一组反转/reverse_list.py
# coding=utf-8

from python.unit3_linklist.part2_hot_problem.list_node import init_list, ListNode, print_linked_list


class SwapPairs:
    def swapPairs(self, head):
        dummyHead = ListNode(0)
        dummyHead.next = head
        temp = dummyHead
        while temp.next and temp.next.next:
            node1 = temp.next
            node2 = temp.next.next
            temp.next = node2
            node1.next = node2.next
            node2.next = node1
            temp = node1
        return dummyHead.next


if __name__ == '__main__':
    # 先构造题目要求的链表,简单构造即可
    nums = [1, 2, 3, 4, 5]
    list = init_list(nums)
    swapPairs = SwapPairs()
    node = swapPairs.swapPairs(list)
    print_linked_list(node)

File: Algorithm/Leetcode/Code_Materials/c_or_cpp/charpter2_reverselist/level2/topic2_2两两一组反转/reverse_list.cpp
#include<stdlib.h>
#include <stdio.h>

#include <stdio.h>
#include "../../list_node.cpp"

  struct ListNode * swapPairs(struct ListNode * head)
{
    struct ListNode *dummyHead = (struct ListNode *)malloc(sizeof(struct ListNode));
    dummyHead->next = head;
    struct ListNode *temp = dummyHead;

    while (temp->next != NULL && temp->next->next != NULL)
    {
        struct ListNode *node1 = temp->next;
        struct ListNode *node2 = temp->next->next;
        temp->next = node2;
        node1->next = node2->next;
        node2->next = node1;
        temp = node1;
    }

    struct ListNode *newHead = dummyHead->next;
    free(dummyHead);
    return newHead;
}

int main()
{
    struct ListNode *p = NULL;
    printf("create list: \t\n");

    int a[] = {1, 2, 3, 4, 5};
    p = initLink(a, 5);
    struct ListNode *q = swapPairs(p);
    printList(q);

    return 0;
}